pymtl3/dsl/Component.py

```diff
@@ -55,10 +61,543 @@ def _construct( s ):
+          del top._dsl.all_adjacency[x]
+
+        # Clean up adjacency at parent
+        if x in parent._dsl.adjacency:
+          for other in parent._dsl.adjacency[x]:
+            # other must be in the dict
+            if other not in removed_connectables:
+              parent._dsl.adjacency[other].remove( x )
+          del parent._dsl.adjacency[x]
+
+      # We don't break nets anymore. Instead, we set the flags to true so
+      # that the next get_xxx_net will immediately recollect nets.
+      top._dsl._has_pending_value_connections = True
+      top._dsl._has_pending_method_connections = True
+
+      # We clean up the connect_order list. If we want to preserve the
+      # original connect order, we can play some other tricks here such as
+      # leaving the strings in there. However I don't think that's clean
+      # enough. Thus I'm just removing them right now.
+      new_connect_order = []
+      for (x, y) in parent._dsl.connect_order:
+        if x not in removed_signals and y not in removed_signals: # TODO method port
+          new_connect_order.append( (x, y) )
+
+      parent._dsl.connect_order = new_connect_order
+
+      return saved_connections
+
+    return _delete_component_internal( top, obj )
+    # import gc
+    # gc.collect() # this takes 0.1 seconds
+
+  #-----------------------------------------------------------------------
+  # Post-elaborate public APIs (can only be called after elaboration)
+  #-----------------------------------------------------------------------
+
+  """ Convenience/utility APIs """
+  def apply( s, *args ):
+
+    if isinstance(args[0], list):
+      assert len(args) == 1
+      for step in args[0]:
+        step( s )
+
+    elif len(args) == 1:
+      assert callable( args[0] )
+      args[0]( s )
+
+  # Simulation related APIs
   def sim_reset( s ):
-    assert s._dsl.elaborate_top is s # assert sim_reset is top
+    s._check_called_at_elaborate_top( "sim_reset" )
 
     s.reset = Bits1( 1 )
     s.tick() # This tick propagates the reset signal
     s.reset = Bits1( 0 )
     s.tick()
+
+  def check( s ):
+    s._check_valid_dsl_code()
+
+  # TODO maybe we should implement these two lock/unlock APIs as passes?
+  # They expose kernel implementation details though ...
+  def lock_in_simulation( s ):
+    s._check_called_at_elaborate_top( "lock_in_simulation" )
+
+    s._dsl.swapped_signals = defaultdict(list)
+    s._dsl.swapped_values  = defaultdict(list)
+
+    def cleanup_connectables( current_obj, host_component ):
+
+      # Deduplicate code. Choose operation based on type of current_obj
+      if isinstance( current_obj, list ):
+        iterable = enumerate( current_obj )
+        is_list = True
+      elif isinstance( current_obj, NamedObject ):
+        iterable = current_obj.__dict__.iteritems()
+        is_list = False
+      else:
+        return
+
+      for i, obj in iterable:
+        if not is_list and i.startswith("_"): # impossible to have tuple
```

Comment:
`is_list` is constant within this for loop so you probably want to move the its check outside `for` to have better performance.